magefile.go

```diff
@@ -1146,15 +1211,388 @@ func packageAgent(platforms []string, packagingFn func()) {
+					log.Printf("platforms include %s", platformString)
+				}
+				filteredPackages[pkgName] = pkgDesc
+				break
+			}
+		}
+	}
+	if mg.Verbose() {
+		log.Printf("filtered packages: %v", filteredPackages)
+	}
+	return filteredPackages
+}
+
+func downloadDRAArtifacts(ctx context.Context, manifestUrl string, downloadDir string, projects ...string) (map[string]tools.Package, error) {
+
+	build, err := manifest.DownloadManifest(manifestUrl)
+	if err != nil {
+		return nil, fmt.Errorf("downloading manifest from %q: %w", manifestUrl, err)
+	}
+
+	// Create a dir with the buildID at <downloadDir>/<buildID>
+	draDownloadDir := filepath.Join(downloadDir, build.BuildID)
+	err = os.MkdirAll(draDownloadDir, 0o770)
+
+	if err != nil {
+		return nil, fmt.Errorf("creating %q directory: %w", err)
+	}
+
+	if mg.Verbose() {
+		log.Printf("url:%q, err %v", manifestUrl, err)
+	}
+
+	// sync access to the downloadedArtifacts map
+	mx := new(sync.Mutex)
+	downloadedArtifacts := map[string]tools.Package{}
+	errGrp, errCtx := errgroup.WithContext(ctx)
+
+	for _, projectName := range projects {
+		project, ok := build.Projects[projectName]
+		if !ok {
+			return nil, fmt.Errorf("project %q not found in manifest at %q", projectName, manifestUrl)
+		}
+
+		if mg.Verbose() {
+			log.Printf("build %q project %s packages: %+v", build.BuildID, projectName, project)
+		}
+		// filter down the packages to the platforms we are building/support
+		filteredPackages := filterPackagesByPlatform(project.Packages)
+		if mg.Verbose() {
+			log.Printf("packages to download: %v", filteredPackages)
+		}
+		for pkgName, pkgDesc := range filteredPackages {
+			downloadFunc := func(pkgName string, pkgDesc tools.Package) func() error {
+				return func() error {
+					artifactDownloadPath := filepath.Join(draDownloadDir, pkgName)
+					err := manifest.DownloadPackage(errCtx, pkgDesc.URL, artifactDownloadPath)
+
+					if err != nil {
+						return fmt.Errorf("downloading %q: %w", pkgName, err)
+					}
+
+					// download the SHA to check integrity
+					artifactSHADownloadPath := filepath.Join(draDownloadDir, pkgName+".sha512")
+					err = manifest.DownloadPackage(errCtx, pkgDesc.ShaURL, artifactSHADownloadPath)
+					if err != nil {
+						return fmt.Errorf("downloading SHA for %q: %w", pkgName, err)
+					}
+
+					err = validateChecksum(sha512.New(), artifactDownloadPath, artifactSHADownloadPath)
+					if err != nil {
+						return fmt.Errorf("validating checksum for %q: %w", pkgName, err)
+					}
+
+					// we should probably validate the signature, it can be done later as we return the package metadata
+					// see https://github.com/elastic/elastic-agent/issues/4445
+
+					mx.Lock()
+					defer mx.Unlock()
+					downloadedArtifacts[artifactDownloadPath] = pkgDesc
+
+					return nil
+				}
+			}(pkgName, pkgDesc)
+
+			errGrp.Go(downloadFunc)
+		}
+	}
+
+	return downloadedArtifacts, errGrp.Wait()
+}
+
+func validateChecksum(hasher hash.Hash, fileToValidate, checksumFile string) error {
```

Comment:
The piece of code that we have in the verifier that validates sha512 is https://github.com/elastic/elastic-agent/blob/d55869484332108b060ebe31bbca779706e77fc4/internal/pkg/agent/application/upgrade/artifact/download/verifier.go#L129-L202

While I am not too comfortable sharing implementations between our build/package system and the code being built/packaged I guess with some light refactor (pulling out the `.sha512` file name instead of assuming that it will always be the same as the name of the file we are validating + '.sha512' extension) I can reuse that part

Reply:
refactor and usage in 19b974aba37e4b45bc9ffb2456ee6690b9398b95